proposals/0000-sum-to-tag.md

```diff
@@ -0,0 +1,165 @@
+the strict scope of this proposal.
+
+If we could only get our hands on the `Int#` tag, then we could calculate
+`tritToInt#` very efficiently. The C-- code would look much like the
+code that would be produced by
+
+```haskell
+tritToInt'# :: (# Int# #) -> Int#
+tritToInt'# (# t #) = t - 1
+```
+
+Unfortunately, that's *not* the sort of code that `tritToInt#` will
+actually produce. Rather, it will produce code like this:
+
+```haskell
+tritToInt''# :: (# Int# #) -> Int#
+tritToInt''# (# 3# #) -> 2#
+tritToInt''# (# 2# #) -> 1#
+tritToInt''# _ -> 0#
+```
+
+This seems pretty silly, considering that the result of `tritToInt#`
+is likely to be fed straight into some numerical computation. There
+will be extra code and unnecessary conditional branches when all that's
+needed is a simple decrement operation.
+
+## Proposed Change Specification
+
+Add a `sumToTag#` primop much like `dataToTag#`. Much like `dataToTag#` can be
+compiled only when the *type constructor* of its argument is known and is an
+algebraic datatype, `sumToTag#` can be compiled only when the *runtime
+representation* of its argument is known and is `'SumRep xs` for some known
+`xs`. `sumToTag#` will return the 0-based index of the sum alternative of its
+argument.
+
+## Examples
+
+```haskell
+sumToTag# ((# | 2# | #) :: (# (# Int, Char #) | Int# | Bool #)) = 1#
+
+-- No need to pin down specific types
+sumToTag# ((# | x | #) :: (# a | x | b :: TYPE 'UnliftedRep #)) = 1#
+```
+
+This will compile, but can never ever be called:
+
+```haskell
+foo :: forall (a :: TYPE ('SumRep '[])). a -> Int#
+foo = sumToTag#
+```
+
+This will compile, but cannot be called with current primitives;
+if that changes in the future, it will always return `0#`.
+
+```haskell
+bar :: forall (a :: TYPE ('SumRep '[ 'IntRep])). a -> Int#
+bar = sumToTag#
+```
+
+## Effect and Interactions
+
+It will be possible to extract tags efficiently. It would be
+beneficial to implement a special rule for `case` of `sumToTag#`, in
+case that should appear in the course of optimization.
+
+```haskell
+case sumToTag# a of
+  0# -> e0
+  1# -> e1
+  _ -> e2
+
+===>
+
+case a of
+  (# _|| #) -> e0
+  (# |_| #) -> e1
+  (# ||_ #) -> e2
+```
+
+One long-term option this primop gives us is to stop returning
+`Int#` from primops whose values represent only two or three
+specific values. For example, we would have the option of
+writing
+
+```haskell
+(==#) :: Int# -> Int# -> (# (##) | (##) #)
+```
+
+which more faithfully represents what the return value can be,
+while still being able to work with the `Int#` representation
+directly using `sumToTag#`. I do *not* propose making such a
+change at this time.
+
+## Costs and Drawbacks
+
+Implementing this primop should be extremely cheap. Implementing
+the special `case` rule will probably not be terribly hard; its
+structure should generally follow that of the rule for `case` of
+`dataToTag#`.
+
+## Alternatives
+
+I do not know of any alternative approach.
```

Comment:
Could this be solved by an optimization in a later phase that detects this pattern, and produces the code that you want?

Reply:
@nomeata that was the first thing I wondered, but @bgamari didn't think it would likely be worth the trouble. I don't personally know what that would look like, but it would presumably require somewhat complicated analysis of almost-common subexpressions in CPP. Even if I knew how to do it, it sounds expensive at compilation time.

Reply:
@nomeata, I've addressed your question in the proposal text.

Reply:
Thanks!